src/Tools/Pagination/Paginator.php

```diff
@@ -171,13 +219,30 @@ private function cloneQuery(Query $query): Query
     /**
      * Determines whether to use an output walker for the query.
      */
-    private function useOutputWalker(Query $query): bool
+    private function useOutputWalker(Query $query, bool $forCountQuery = false): bool
     {
-        if ($this->useOutputWalkers === null) {
-            return (bool) $query->getHint(Query::HINT_CUSTOM_OUTPUT_WALKER) === false;
+        if ($this->useOutputWalkers !== null) {
+            return $this->useOutputWalkers;
         }
 
-        return $this->useOutputWalkers;
+        // When a custom output walker already present, then do not use the Paginator's.
+        if ($query->getHint(Query::HINT_CUSTOM_OUTPUT_WALKER) !== false) {
+            return false;
+        }
+
+        // When not joining onto *ToMany relations, then do not use the more complex CountOutputWalker.
+        // phpcs:ignore SlevomatCodingStandard.ControlStructures.UselessIfConditionWithReturn.UselessIfCondition
```

Comment:
Note: I think it's better not to enforce the failing coding standard on that if-condition because otherwise it's a multiline `return` abomination.

Reply:
It's unclear to me why the `return` abomination would be worse than the `if` abomination.

Reply:
1. The way I see it is that this function is a set of if-statements that each check for a very individual use case. If none of the statements "wants to make a decision", then the default "return true;" is returned. 
2. The if-condition for the new if-statement is quite elaborate, so I didn't want it to be the last statement of the function.
3. The next dev that wants to add more if-statements to this function, would have to "un-return" my if-statement. This argument is my common argument for all phpstan rules that enforce "return statement code style" rules, which often in my mind are "false positives".

These points are semantics + I'm the "guest" here, so I've merged the last two return statements, as per phpstan detection.